### PR TITLE
Local git

### PR DESCRIPTION
**Pull request by jimaek**

### CODE REVIEW

**Comment by megawac** [comment]
Ready for :eyes: /cc @UnbounDev

File: config/config.template.js [hunk 1]
```diff
@@ -19,10 +19,20 @@ module.exports = {
...
+    },
+    'google': {
+      minute: 3
+    },
+    'cdnjs': {
+      minute: 1,
+      gitPath: '/home/gyeates/code/cdnjs'
+    },
+    'jsdelivr': {
+      minute: 1,
```
Comment (megawac) [inline config/config.template.js]: btw, `minute: x` doesn't do anything at the moment

File: tasks/cdnjs.js [hunk 2]
```diff
@@ -1,166 +1,120 @@
...
+function parse(projects, cb) {
+  var ret = [];
 
-  async.eachLimit(files, 4, function(file, cb) {
+  async.eachOf(projects, function(versions, projectName, cb) {
+    var proj = {
+      name: projectName,
+      versions: [],
+      assets: {}, // version -> assets
+      zip: projectName + '.zip'
```
Comment (megawac) [inline tasks/cdnjs.js]: Should this be on the `cdnjs` configs?

**Comment by jimaek** [comment]
Can you explain to me how it works? Its just a cli tool than I need to run on a cronjob?
How does it store the data and connects to the api frontend?

**Comment by megawac** [comment]
It doesn't change how `api-sync` is called. It just changes how it works under the hood. As we've previously discussed, it requires a local git folder for both `cdnjs` and `jsdelivr`. From there it does a `git rebase` against origin for both repos and syncs as normal.

Based on how `api-sync` is coded it looks like it should be a dameon process but it has a couple things that could use further work (as I mentioned in a comment the `minute` option is currently not implemented to do anything).

**Comment by jimaek** [comment]
We need to update the documentation explaining how it works exactly. How to setup from 0 and how to maintain it.
In the current form as I understand it basically does nothing. A daemon with no internal crontab.

**Comment by megawac** [comment]
It looks like it resyncs itself every 30 minutes https://github.com/jsdelivr/api-sync/blob/master/serve.js#L123-L132 currently

**Comment by megawac** [comment]
Again, that should probably be changed to respect a `sync` timer on a per cdn basis

**Comment by jimaek** [comment]
Since its local we can set it to 30 minutes, same as libgrabber. Right?
What about bootstrapcdn? How are you getting that data?

**Comment by megawac** [comment]
Lets move that discussion to #38 @jimaek. Bootstrap cdn is as it was before, through GH api

**Comment by megawac** [comment]
Also, why do you want it so slow, if its local you can do it as often as you want without expecting any issues

**Comment by megawac** [comment]
This will require a further change to https://github.com/jsdelivr/api/blob/master/tasks/sync.js as `cdnjs` and `jsdelivr` will no longer generate etags for the project. I especially question the wisdom of creating a data folder with a file for each project; this is very inefficent for the first sync as a file for each project has to be fetched.

I still question the wisdom of the current approach and think two approaches would be better:
1) make api-sync use a conventional database (couchdb, aerospike, whatever) and be responsible for writing/updating entries. api would connect and serve items from this database and not be responsible for syncing in any way.
2) make api-sync have an end point for each cdn (bootstrap, jsdelivr, google, etc) which returns a json string of all items currently hosted. This would increase the network load but require less logic on `jsdelivr/api` in determining individual projects which have changed during syncing --- it can instead just merge the json with the current database. 

------------------------------------

TLDR: minimize the sync logic on `api`, make `api-sync` the sole project responsible for syncing; rethink communication format between `api` and `api-sync`

/cc @jimaek @UnbounDev

**Comment by UnbounDev** [comment]
tldr; read this! It's far easier than digging through the code.

### How `api-sync`, syncs

By starting the api-sync application you're initializing a long running node application that will perform projects syncs at it's own discretion, it is not a cron job.

1. A sync of any project begins by hitting the [`triggerTask`](https://github.com/jsdelivr/api-sync/blob/master/serve.js#L139) function, this is triggered for every project at 30 minute intervals and in real time for the `jsdelivr` repo (via the [webhook express route](https://github.com/jsdelivr/api-sync/blob/master/serve.js#L77-L91)).
2. When a sync is triggered the application grabs repo data from github (using the github api in the old case, and via the use of a local git repo in Graeme's implementation), this data is then parsed into a formate `api` knows how to read and stored in individual folders for retrieval by the jsdelivr api; further, a 'master' json file is created per repo that includes a map of the project names retrieved by each sync along w/ the corresponding eTag for each project. 
  - **The reason for this storage structure is that we are severely limited in computation power by our appfog instances,** in order to perform a sync from api-sync to api w/o crashing the application we must be able to consume a single project at a time.

### How `api` consumes `api-sync` data

 `api` sustains it's own sync cycle on 10 minute intervals, during a sync it performs the following steps for each CDN we track:

1. Pulls the master json file for each repo and compares the project eTag's to it's own project->eTag list.
2. For each project whose eTag does not match api's own record, api requests the json file for that project from api-sync and updates it's internal eTag/project record.

### Responses to various points on this thread

 - I do agree w/ @megawac in his suggestion that api-sync make use of a relational database (I prefer postgres, personally), but last I heard we were still waiting on project sponsorship before enhancing the platform resources.
 - ^ in response to that, it's $15 a month on AWS to get a small postgres instance up, we're not concerned w/ the size of the DB in this case as it is only in use as an intermediary between api-sync and api (never being subject to end-user requests).
 - @megawac: regarding the eTag propagation, while this is primarily used to maximize the efficiency w/ which api performs its own sync cycle, keep in mind that this implementation also grants us an appropriate 'last updated' time for that project which is a feature of api v2.

**Comment by megawac** [comment]
Yeah, I'm familiar with how sync works and I hate it. I would like to see the logic reduced to something super simple (pseudo code below) instead of having significant logic.

```
for cdn in cdns:
    data = request.get(<api-sync-service>/cdn)
    database[cdn].updateDatabase(data)
```

> @megawac: regarding the eTag propagation, while this is primarily used to maximize the efficiency w/ which api performs its own sync cycle, keep in mind that this implementation also grants us an appropriate 'last updated' time for that project which is a feature of api v2.

It would be better to add this to the json in that case. Best would be to actually check the timestamp of the commit that updated the project on the jsdelivr repo. The eTag efficency would be negligible if we went with either of my suggestions and avoided a unique file per project (**thus reducing >4k requests to 5 on updates**)

> ^ in response to that, it's $15 a month on AWS to get a small postgres instance up, we're not concerned w/ the size of the DB in this case as it is only in use as an intermediary between api-sync and api (never being subject to end-user requests).

Using a master json document per CDN in the format API builds its mem database would avoid the requirement of getting a server that supports a relational database.

**Comment by UnbounDev** [comment]
The point I'd like to stress here is that the current sync structure is designed to mitigate application crashes on appfog due to exceeded the machines memory limit - those boxes are tiny.

Details:
- It is not feasible to pull a master json file that contains all CDN projects down to tbe api on each api sync cycle, this was the original sync structure and it caused the api to crash on multiple occasions (always w/ a memory exceeded error from appfog). The (relatively) excessive memory usage was caused by the api having to keep both the current api db in memory as well as the entire newly pulled CDN data from api-sync (note that the real culprit here is the mathjax library, damn thing's api metadata is over the 50mb mark).
- It is not >4k requests per update, it's >4k requests on application start - after that api only requests projects that have *changed* + a request for the CDN manifest from api-sync (you can easily check this by running api locally and set the sync interval to something like a minute), it seems perfectly reasonable to have a high startup cost given the massive performance increase on subsequent sync cycles. **Again, I implemented that strategy b/c the api was consistently crashing on appfog, and it has worked thus far.**
- I agree about the timestamp/commit bit and including that in the json for the project, if your implementation is able to determine that from the local git logs then absolutely (I only used the eTags b/c they were convenient).

**Comment by megawac** [comment]
> It is not >4k requests per update, it's >4k requests on application start - after that api only requests projects that have changed

It will still have to initiate an HTTP request to check the etag for each project

> I agree about the timestamp/commit bit and including that in the json for the project, if your implementation is able to determine that from the local git logs then absolutely (I only used the eTags b/c they were convenient).

Would be super easy for any project using git but may be difficult with the google cdn

**Comment by UnbounDev** [comment]
> Correct me if I'm wrong...

Ah! Happily corrected :) and to be clear I'm talking about api consuming data from api-sync. The first request api makes is for a small manifest file for the CDN, it then compares the projectName->eTag list in that file w/ it's own, it will only request project updates if the eTag has changed.

Yea the google CDN is so small it's fine to request each time, believe api-sync gen's a new 'eTag' for those libraries on each sync to force api into an update.

**Comment by jimaek** [comment]
btw just to make sure its clear to everyone. If we switch api-sync to local scanning then it will be hosted on a VPS and not appfog. So RAM should not be a huge concern anymore (for `api-sync`)
But `api` frontend will still remain on appfog for now.

I also have an RDS mysql DB that you can use if you need but I prefer to avoid to purchase or use any other local/remote DBs.

**Comment by UnbounDev** [comment]
@megawac are we good w/ replicating the current structure using the local git commit tag in place of eTags? If so the rest of the system should just work and all features will remain as they are today for api v1 and v2.

**Comment by megawac** [comment]
What do you mean by that? It should be creating the same json under data/
if thats what you mean

**Comment by UnbounDev** [comment]
@megawac yes that's what I meant, if it creates the json under `data/` then it should work w/o any modifications to `api`. 

I'm trying to run this locally though and it looks like I have to init a git repo for both jsdelivr and cdnjs? Can we change this to clone the repos on startup if they don't exist to make deployment easier (and to enforce a standard location for the repositories)? Preferably into a sibling directory to `api-sync`:
```
\home
  \api-sync
  \jsdelivr
  \cdnjs
```

**Comment by UnbounDev** [comment]
Also - this isn't creating all of the appropriate json files, for instance, syncing jsdelivr content should result in:

```
/data
  /jsdelivr.json
  /jsdelivr
    /projects....
```

the `jsdelivr.json` is missing, it should be a json file w/ the following structure:

```
[
  {
    "path": "<ProjectName>",
    "etag": "\"<ETag or CommitHash>\"",
    "sha": "<CommitHash>"
  },
  ...
]
```

If you run the sync for `bootstrap-cdn` you'll see the appropriate structure for that repo, we need to mimic that structure for `api` to continue working as-is.

**Comment by megawac** [comment]
I know its not creating the etag json file. Its one of the reasons I
brought up the synchronization strategies to determine if we want to change
them from the current strategy to a new strategy.

I'd rather let the user clone the repo & set the path -- leaving the remote
generic. Currently there should be a clear error if you didn't set the
gitPath for jsdelivr/cdnjs

**Comment by UnbounDev** [comment]
Alright - so long as we're limited by the instance size I don't see the reason in changing synchronization strategies for `api`, but that shouldn't matter if we can construct the etag file based on data from your local-git implementation.
